### ISSUE TITLE
External opening books & other features

### ISSUE THREAD

**Issue by ghost** [Psyyke/A.C.A.S#33]
It would be very useful if it were possible to configure an arrow color for a strong engine, for example Stockfish 16, and an arrow of another color for a more "human" engine like Maia 1900, to simulate a human when the strongest engine's move is very weird, counterintuitive or impractical.

Just a suggestion

**Comment by Hakorr** [comment]
This is already possible, no?

**Comment by mattt720** [comment]
> This is already possible, no?

Yes, inherantly.

I'm a big fan of what you've made, so far I just have some questions;
1. Has your opinion changed of Lc0 since implementing Maia-1100? What do you think of it in terms of human-likeness?
2. Is there any combination of engines / settings you've found particularly favourable in terms of human-like performance?
3. Are there any other engines you might plan on adding in the future?

Thank you again for what you do, it's a great learning tool for someone like me who struggles to map things out in my head.

**Comment by Hakorr** [comment]
1. It's more realistic than Stockfish, however relying solely on engine moves regardless of Lc0 or Stockfish is detectable as far as I know. There doesn't seem to be an engine out there which can imitate a human perfectly. Maybe with heavy modifications and making sure correct move times are followed, one could automate a certain ELO playstyle quite accurately.
2. There is no config which wouldn't get you banned if you used the moves the engine suggested too much. Just keep the engine's skill levels low and keep your playing style consistent across games.
3. I'll add more engines as I discover them. They need to be built for WASM/JavaScript, be open source and licensed in a way I can use them. For example I could add the Chess.com Komodo engine and such however that would break their license, so I won't.

> Thank you again for what you do, it's a great learning tool for someone like me who struggles to map things out in my head.

Thanks, do you have any feature suggestions that would help with your learning?

**Comment by mattt720** [comment]
> 1. It's more realistic than Stockfish, however relying solely on engine moves regardless of Lc0 or Stockfish is detectable as far as I know. There doesn't seem to be an engine out there which can imitate a human perfectly. Maybe with heavy modifications and making sure correct move times are followed, one could automate a certain ELO playstyle quite accurately.
> 2. There is no config which wouldn't get you banned if you used the moves the engine suggested too much. Just keep the engine's skill levels low and keep your playing style consistent across games.
> 3. I'll add more engines as I discover them. They need to be built for WASM/JavaScript, be open source and licensed in a way I can use them. For example I could add the Chess.com Komodo engine and such however that would break their license, so I won't.
> 
> > Thank you again for what you do, it's a great learning tool for someone like me who struggles to map things out in my head.
> 
> Thanks, do you have any feature suggestions that would help with your learning?

I agree in relying on any engine is pointless, but rather being able to play the most human-like engine infinitely helps, given stockfish's repitoire of strange non-human moves and pointless blunders.

In terms of features, I have nothing specific but can happily try contribute some value:
- An [opening / endgame] feature that could highlight how to, according to chess literature, respond to detected [openings / endgames] or execute them when specific positions are aligned.
- Blunder detection (e.g. highlight arrow red if blunder) & feedback or Move explaination (not sure how these would be done - perhaps feeding the FEN to something like an LLM or via API?): A low-level player like me would love to further understand the moves of higher up players, rather than having to figure them out on my own board. The blunder detection would be good so users can know what engine suggested _non_-top move options would and wouldn't be blunders?
- Heatmap feature or pattern recognition feature? Heatmap could be post-game whilst pattern recognition could be real-time in identifying known piece formations and possibly their pros & cons; AI could certainly help with this, though perhaps not real-time...
- Lastly, top-engine / high-engine move detection. I'd like to know which players require reporting easily given I'm unsure how to distinguish a cheater who mixes in human moves with engine ones. Repeating highlighted top/high-engine moves could go towards a counter, or simply notify the player.

I acknowledge and apologise for how large / difficult they may be to implement as I'm unsure of your restrictions given I'm not an experienced coder either. Hope these might be of some use or inspiration and would definitely help beginners such as myself.

**Comment by trashpro7** [comment]
> 1. It's more realistic than Stockfish, however relying solely on engine moves regardless of Lc0 or Stockfish is detectable as far as I know. There doesn't seem to be an engine out there which can imitate a human perfectly. Maybe with heavy modifications and making sure correct move times are followed, one could automate a certain ELO playstyle quite accurately.
> 2. There is no config which wouldn't get you banned if you used the moves the engine suggested too much. Just keep the engine's skill levels low and keep your playing style consistent across games.
> 3. I'll add more engines as I discover them. They need to be built for WASM/JavaScript, be open source and licensed in a way I can use them. For example I could add the Chess.com Komodo engine and such however that would break their license, so I won't.
> 
> > Thank you again for what you do, it's a great learning tool for someone like me who struggles to map things out in my head.
> 
> Thanks, do you have any feature suggestions that would help with your learning?

The best way to learn chess is with pattern recognition, and playing with Maia 1900 has helped me a lot. The developers have released "Maia Individual" (https://github.com/CSSLab/maia-individual) which allows you to train Maia on games played by a specific player, and he can guess 65% of the moves that particular player would make, while playing like a human. It would be very useful if it were possible to load lc0 weights in A.C.A.S from the computer, but I don't know how challenging that would be. With that, it would be possible to train Maia on games from a specific rating range, or from a specific Grandmaster, and then load it in A.C.A.S as external assistance.

**Comment by Hakorr** [comment]
>  With that, it would be possible to train Maia (locally) on games from a specific rating range, or from a specific Grandmaster, and then load the weight file in A.C.A.S as external assistance.

Very good idea, I like that it could mimic the player's own playing style.

**Comment by trashpro7** [comment]
> > With that, it would be possible to train Maia (locally) on games from a specific rating range, or from a specific Grandmaster, and then load the weight file in A.C.A.S as external assistance.
> 
> Very good idea, I like that it could mimic the player's own playing style.

Thank you. Do you have any plans to add what I suggested, about "Maia Individual", or is it not a realistic goal? I intend to close the issue. Although the original issue has already been answered, I thought it would be better to present my proposal in this same issue, instead of opening another one, so as not to be annoying.

**Comment by Hakorr** [comment]
Your original suggestion is a bit harder to add so it's a feature for the future, I didn't want to close this issue since it has great conversation and ideas I want to come back to.

The idea of a minimum move time is good, and easy to add. I'll probably add it along the next update.

**Comment by trashpro7** [comment]
I have a new suggestion

I know that some engines use some sort of external opening books (i.e. the engine itself doesn't make the decisions). Can you add them? https://www.chessprogramming.org/Opening_Book

https://github.com/gmcheems-org/free-opening-books

And I know the App "Lucas Chess" uses a GM book called "GM Openings". It would be interesting if you could add it as well!

This could make the engine play ore human like

Btw, there are a few tools for creating opening books. Bookbuilder, Bookup, Chess Opening Wizard (COW) and Baksia Gui, etc. How difficult would it be to add an option to add an opening book in bin format ourselves?

**Comment by Copameds28** [comment]
> I have a new suggestion
> 
> I know that some engines use some sort of external opening books (i.e. the engine itself doesn't make the decisions). Can you add them? https://www.chessprogramming.org/Opening_Book
> 
> https://github.com/gmcheems-org/free-opening-books
> 
> And I know the App "Lucas Chess" uses a GM book called "GM Openings". It would be interesting if you could add it as well!
> 
> This could make the engine play more human-like
> 
> Btw, there are a few tools for creating opening books. Bookbuilder, Bookup, Chess Opening Wizard (COW) and Baksia Gui, etc. How difficult would it be to add an option to add an opening book in bin format ourselves?

It would be really great if you could add this feature.

**Comment by ghost** [comment]
> > it makes more sense for it to be in the web GUI.
> 
> Yup. Anyhow, I'll make this feature during summer. Feel free to add any further details before that if you feel like.

Hey there. Just letting you know that Maia-3 has been released, if you're interested.

https://github.com/CSSLab/maia3

Unlike the legacy version which ranged from 1100 to 1900, it seems that Maia 3 ranges from 600 to 2600 in rating, with an accuracy of 62.5% in mimicking human moves within each rating range.

**Comment by Hakorr** [comment]
> Hey there. Just letting you know that Maia-3 has been released, if you're interested.

Hey, I added Maia-3 when it was released (well, I took it from another project of CSSLab, so weeeeks ago before the official release date). If you're using A.C.A.S locally, please update the files by cloning the repo again. It should be visible on the GUI.

But thanks for letting me know, there's actually some settings like Temperature and TopP that A.C.A.S doesn't support at the moment, so I should add them. There are also different models to choose from so I need to look into that.

**Comment by Copameds28** [comment]
@Hakorr, I recently found a great website, [take a look at this](https://chessfish.io/). I really like the evaluation feature here.

I found it here in this [subreddit](https://www.reddit.com/r/SideProject/comments/1jrjbl3/i_made_a_chess_app_that_shows_you_every_move/)

**Comment by Hakorr** [comment]
> I recently found a great website, [take a look at this](https://chessfish.io/). I really like the evaluation feature here.

It's certainly nice. You can already get similar results by making two engines on A.C.A.S and activating "moves on demand" on one. Is that not good enough for that specific feature?

But the eval texts and showing the move's intention (?) would need to be added. Good suggestion.

Also a side note, there's no reject cookies button on the site, weird. Violates GDPR and is basically illegal in sites operating in EU.
